File: Bit Manipulation/Change Bits/change_bits.lua
local lines = {}
for line in io.lines() do lines[#lines + 1] = line end
local N, Q = string.match(lines[1], "(%d+) (%d+)")

lines[2]:reverse()
local A = {} 
lines[2]:gsub('%d', function(n) A[#A + 1] = n end)

lines[3]:reverse()
local B = {}
lines[3]:gsub('%d', function(n) B[#B + 1] = n end)

local C = {}

local function set(signal, index, value)
	if signal == 'a' then
		A[index] = value
	else 
		B[index] = value
	end
end

local function get(index)
	local carry = 0
	for i=1, N do
		local sum = A[i] + B[i] + carry
		if sum == 3 then
			C[i] = '1'
			carry = 1
		elseif sum == 2 then
			C[i] = '0'
			carry = 1
		elseif sum == 1 then
			C[i] = '1'
			carry = 0
		else
			C[i] = '0'
			carry = 0
		end
	end

	if (carry == 1) then
		C[N + 1] = '1'
	else 
		C[N + 1] = '0'
	end

	return C[index]
end

for i=4,#lines do
	local parameters = {}
	lines[i]:gsub('%d+', function(n) parameters[#parameters + 1] = n end)
	-- for i,v in ipairs(parameters) do
	-- 	print(i,v + 1)
	-- end
	if string.sub(lines[i], 1, 3) == "set" then
		set(lines[i]:sub(5, 5), parameters[1] + 1, parameters[2])
	else
		io.write(get(parameters[1] + 1))
	end
end

File: Bit Manipulation/Change Bits/change_bits.c
#include <stdio.h>

void add(unsigned char A[], unsigned char B[], unsigned char C[], unsigned char carry_table, int N, int last_carry_index);

int main()
{
	int N, Q;
	scanf("%d%d", &N, &Q);
    unsigned char A[N + 1], B[N + 1], C[N + 2], carry_table[N + 2];
    scanf("%s%s", A, B);
    char change = 'Y', signal, cmd[10], v;
    int index;

    int last_carry_index = 0;

    for(int i = 0; i < Q; i++){
    	scanf("%s %d", cmd, &index);
    	if(cmd[0] == 's'){
    		scanf(" %c", &v);
    		signal = cmd[4];
            if (signal == 'a'){
                if (A[N - 1 - index] != v){
                    A[N - 1 - index] = v;
                    change = 'Y';
                    if((N - 1 - index) > last_carry_index)
                    	last_carry_index = N - 1 - index;
                }
            }else{
                if (B[N - 1 - index] != v){
                    B[N - 1 - index] = v;
                    change = 'Y';
                    if((N - 1 - index) > last_carry_index)
                    	last_carry_index = N - 1 - index;
                }
            }
//    		(signal == 'a') ? (A[N - 1 - index] = v) : (B[N - 1 - index] = v);
    	}else{
    		if (change == 'Y'){
    			add(A, B, C, N);
    			change = 'N';
    		}
    		printf("%c", C[N - index]);
    	}
    }
    return 0;
}

void add(unsigned char A[], unsigned char B[], unsigned char C[], unsigned char carry_table, int N, int last_carry_index){
	unsigned char carry = carry_table[last_carry_index];
	for (int i = last_carry_index; i < N; ++i)
	{
		unsigned char sum = A[N - 1 - i] + B[N - 1 - i] + carry - '0' - '0';
		C[N - i] = (sum == 3 || sum == 1) + '0';
		carry_table[N - i] = carry = (sum > 1);
	}
	C[0] = carry + '0';
}